libraries/AP_LocationDB/AP_LocationDB.h

```diff
@@ -0,0 +1,139 @@
+    friend class AP_OABendyRuler;
+
+    AP_LocationDB();
+
+    CLASS_NO_COPY(AP_LocationDB);
+
+    static AP_LocationDB *get_singleton() {
+        return _singleton;
+    }
+
+    void init();
+    void update();
+
+    enum class KeyDomain : uint8_t {
+        // 0 should not be used
+        MAVLINK = 1U,
+#if AP_LOCATIONDB_KEYDOMAIN_ADSB_ENABLED
+        ADSB = 2U,
+#endif
+#if AP_LOCATIONDB_KEYDOMAIN_SCRIPTING_ENABLED
+        SCRIPTING = 3U,
+#endif
+    };
+
+    bool add_item(AP_LocationDB_Item item); // add item to the location database
+    bool get_item(const uint32_t key, AP_LocationDB_Item &ret); // get item with given key from location database
+    bool update_item(const uint32_t key, AP_LocationDB_Item &new_item); // update item with given key in location database
+    bool remove_item(const uint32_t key); // remove item with given key from location database
+    bool healthy() const { return _items != nullptr; } // returns true when memory is allocated to store database items
+    bool is_full() const { return _capacity == _size; }; // returns true when database is full
+    void clear(); // clear location database, i.e., remove all items
+    uint16_t size() const { return _size; } // return number of items present in location database
+    uint16_t capacity() const { return _capacity; } // return the capacity of location database
+    bool item_exists(const uint32_t key); // return true when item with given key exists in the location database
+    
+    static bool is_valid_key(const uint32_t key); // return true if a given key is a valid location database key
+    static uint32_t construct_key_mavlink(const uint8_t sysid, const uint8_t compid, const uint8_t msgid); // helper method to construct location database key for a mavlink item
+#if AP_LOCATIONDB_KEYDOMAIN_ADSB_ENABLED
+    static uint32_t construct_key_adsb(const uint32_t icao); // helper method to construct location database key for a adsb item
+#endif
+#if AP_LOCATIONDB_KEYDOMAIN_SCRIPTING_ENABLED
+    static uint32_t construct_key_scripting(const uint32_t subkey); // helper method to construct location database key for a scripting item
+#endif
+    static uint8_t short_mav_msg_id(const uint32_t msgid); // shortened 8 bit message id for a mavlink message
+
+    // parameter var table
+    static const struct AP_Param::GroupInfo var_info[];
+
+    // parameters
+    AP_Int16        _database_capacity_param;                   // db capacity
+    AP_Int8         _default_flags;                             // default flags on new database item
+    AP_Int16        _item_timeout;                              // timeout for db item
+    AP_Int16        _inc_radius;                                // inclusion radius
+
+private:
+
+    uint16_t _capacity;
+    uint16_t _size;
+    AP_LocationDB_Item* _items;
+    HAL_Semaphore db_sem;
+
+    bool get_item_index(const uint32_t key, uint16_t &index); // get index of a database item with given key
```

Comment:
```suggestion
    bool get_item_index(const uint32_t key, uint16_t &index) const; // get index of a database item with given key
```

Reply:
Same goes here. It needs to acquire a semaphore internally.

Reply:
This is a good application for the `mutable` keyword, but it's not allowed in AP. Leave it as-is.